src/main/java/net/imagej/ops/segment/hough/HoughCircleDetectorLocalExtremaOp.java

```diff
@@ -0,0 +1,216 @@
+import net.imglib2.Sampler;
+import net.imglib2.algorithm.localextrema.LocalExtrema;
+import net.imglib2.algorithm.localextrema.LocalExtrema.LocalNeighborhoodCheck;
+import net.imglib2.algorithm.localextrema.RefinedPeak;
+import net.imglib2.algorithm.localextrema.SubpixelLocalization;
+import net.imglib2.algorithm.neighborhood.Neighborhood;
+import net.imglib2.type.NativeType;
+import net.imglib2.type.numeric.RealType;
+import net.imglib2.util.Util;
+
+/**
+ * A Hough circle detector based on finding local extrema in the vote image,
+ * followed by non-maxima suppression.
+ * <p>
+ * This detector offers typically a faster processing than its DoG-based
+ * counterpart, at the cost of accuracy and robustness.
+ * 
+ * @author Jean-Yves Tinevez
+ *
+ * @param <T>
+ *            the type of the source vote image.
+ */
+@Plugin( type = HoughCircleDetectorOp.class )
+public class HoughCircleDetectorLocalExtremaOp< T extends RealType< T > & NativeType< T > >
+		extends AbstractUnaryFunctionOp< RandomAccessibleInterval< T >, List< HoughCircle > >
+		implements Cancelable
+{
+
+	@Parameter
+	private ThreadService threadService;
+
+	@Parameter( required = true, min = "1" )
+	private double minRadius;
+
+	@Parameter( required = true, min = "1" )
+	private double stepRadius;
+
+	@Parameter( required = false, min = "0.1" )
+	private double sensitivity = 20.;
+
+	@Override
+	public List< HoughCircle > calculate( final RandomAccessibleInterval< T > input )
+	{
+		final int numDimensions = input.numDimensions();
+
+		/*
+		 * Find local extrema.
+		 */
+
+		final double threshold = 2. * Math.PI * minRadius / sensitivity;
+		final T t = Util.getTypeFromInterval( input ).createVariable();
+		t.setReal( threshold );
+		final LocalNeighborhoodCheck< Circle, T > maximumCheck = new LocalNeighborhoodCheck< Circle, T >()
+		{
+			@Override
+			public < C extends Localizable & Sampler< T > > Circle check( final C center, final Neighborhood< T > neighborhood )
+			{
+				final T c = center.get();
+				if ( t.compareTo( c ) > 0 )
+					return null;
+
+				for ( final T t : neighborhood )
+					if ( t.compareTo( c ) > 0 )
+						return null;
+
+				final double val = c.getRealDouble();
+				final double radius = minRadius + ( center.getDoublePosition( numDimensions - 1 ) ) * stepRadius;
+				final double ls = 2. * Math.PI * radius / val;
+				return new Circle( center, radius, ls );
+			}
+		};
+		final ArrayList< Circle > peaks = LocalExtrema.findLocalExtrema( input, maximumCheck, threadService.getExecutorService() );
+
+		if ( isCanceled() )
+			return Collections.emptyList();
+
+		/*
+		 * Non-maxima suppression.
+		 * 
+		 * Rule: when one circle has a center inside one another, we discard the
+		 * one with the highest sensitivity.
+		 */
+
+		// Sort by ascending sensitivity.
+		Collections.sort( peaks );
+
+		final List< Circle > retained = new ArrayList<>();
+		NEXT_CIRCLE: for ( final Circle tested : peaks )
+		{
+			for ( final Circle kept : retained )
```

Comment:
Can you explain to me the purpose of removing any circles inside circles? I do not understand this.

Reply:
This is the non-maxima suppression part. 
The local extrema-based detector directly looks for local maxima in the vote image. 

The vote image might be 'noisy': several local maxima might correspond to the same true circle but be separated by a couple of pixels. This will results in having two or more circles detected very close to each other with a very similar radius. 

Non-maxima suppression is an ad-hoc workaround for these artefacts. You order all the circles by some quality factor (we use the sensitivity) and if two circles are found too close (in my case, one inside one another) you just keep the one with the best quality (lower sensitivity).

This is not always desirable if you know that some smaller circles are supposed to be inside larger one. In that case you have to use the other detector.

Reply:
So if we use the DOG Detector then we will find smaller circles inside larger ones?